src/types.rs

```diff
@@ -422,13 +446,75 @@ impl<N: Name> Type<N> {
             }
             Type::Variable(v) => {
                 *self = ctx
-                    .substitution
+                    .cache
                     .get(&v)
+                    .or_else(|| ctx.substitution.get(&v))
                     .map(|tp| tp.apply(ctx))
                     .unwrap_or_else(|| Type::Variable(v));
             }
         }
     }
+    /// Applies the type in a [`Context`], performing path compression.
+    /// 
+    /// Path compression is a technique commonly used in [Union-Find data structures]. We apply it
+    /// here so that whenever a chain of substitutions is traversed, each variable is updated to
+    /// point to its ultimate value. For example, the chain:
+    ///
+    /// `t0 ↦ t1`, `t1 ↦ t2`, and `t2 ↦ int`
+    ///
+    /// becomes
+    /// 
+    /// `t0 ↦ int`, `t1 ↦ int`, and `t2 ↦ int`
+    ///
+    /// Rather than updating the actual mappings, [`Context`] maintains a `cache` of these
+    /// compressed mappings.
+    ///
+    /// [Union-Find data structure]: https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+    /// [`Context`]: struct.Context.html
+    pub fn apply_compress(&self, ctx: &mut Context<N>) -> Type<N> {
```

Comment:
So it seems like this is strictly better than `apply`. Could we make Context use interior mutability for `cache`, and have regular apply/apply_mut perform path compression? Given this should have no affect on behavior, the interior mutability shouldn’t be very controversial.. it is a `cache`, after all

Reply:
The latest commits make this change. Path compression is now part of the default `apply` and `apply_mut`, and I removed the `_compress` methods.